Restructuring to allow preprocessing, tracking and rendering to be compiled separately

sotpapathe opened the pull request:
Changed the project structure so that preprocessing, tracking and rendering are split into hpp and cpp files. This should make code separation cleaner and allow faster compilation for incremental changes.

### Preprocessing
Split into an hpp and cpp and added explicit instantiation of template function `vertex2normalKernel()`. The template was for a `bool` so all possible functions have been instantiated.

### Tracking
Just split into an hpp and cpp. No notable changes.

### Structural changes
- Moved the `SE_FIELD_TYPE` `typedef` from `DenseSLAMSystem.h` to `volume_traits.hpp ` so that it will be available to all headers (rendering.hpp etc).
- Moved the `bfusion` and `kfusion` folders from `src` to `include`.
- Some bug fixes and code style improvements.

### Rendering
Split into hpp and cpp. Two functions were split fine (`renderDepthKernel()` and `renderTrackKernel()`) but the other two (`raycastKernel()` and `renderVolumeKernel()`) were templated andI left them in the header for now. I am not particularly happy about this but maybe it would be better to use the new file structure as early as possible and make rendering cleaner later.

### Benchmarks
I run the generated benchmarks for both the SDF and occupancy mapping representations for the old and new code. I don't think the changes had an impact in performance, although the benchmarking wasn't that thorough. Each benchmark was run 10 times and the results are below. What is strange is that in both the new and old code the time to complete the SDF benchmark varied a lot. No clue what might be causing this at the moment.

New code
```bash
# SDF TUM
# Benchmarking './build/se_apps/se-denseslam-sdf-benchmark -i /home/srl/Documents/Datasets/TUM_RGBD/rgbd_dataset_freiburg1_desk_mrcnn/scene.raw -s 6 -p 0.5,0.5,0.5 -z 4 -c 2 -r 1 -k 525,525,319.5,239.5'
# for 10 iterations
-----------------------------------
Mean:                 22.928244 s
Standard deviation:   12.319538 s
Min:                  8.892734 s
Max:                  38.106689 s
Total time:           229.282444 s

# OFUSION TUM
# Benchmarking './build/se_apps/se-denseslam-ofusion-benchmark -i /home/srl/Documents/Datasets/TUM_RGBD/rgbd_dataset_freiburg1_desk_mrcnn/scene.raw -s 6 -p 0.5,0.5,0.5 -z 4 -c 2 -r 1 -k 525,525,319.5,239.5'
# for 10 iterations
-----------------------------------
Mean:                 23.038625 s
Standard deviation:   0.911637 s
Min:                  20.769309 s
Max:                  24.296465 s
Total time:           230.386250 s
```

Old code
```bash
# SDF TUM
# Benchmarking './build/se_apps/se-denseslam-sdf-benchmark -i /home/srl/Documents/Datasets/TUM_RGBD/rgbd_dataset_freiburg1_desk_mrcnn/scene.raw -s 6 -p 0.5,0.5,0.5 -z 4 -c 2 -r 1 -k 525,525,319.5,239.5'
# for 10 iterations
-----------------------------------
Mean:                 22.968320 s
Standard deviation:   12.656391 s
Min:                  8.807932 s
Max:                  39.406392 s
Total time:           229.683196 s

# OFUSION TUM
# Benchmarking './build/se_apps/se-denseslam-ofusion-benchmark -i /home/srl/Documents/Datasets/TUM_RGBD/rgbd_dataset_freiburg1_desk_mrcnn/scene.raw -s 6 -p 0.5,0.5,0.5 -z 4 -c 2 -r 1 -k 525,525,319.5,239.5'
# for 10 iterations
-----------------------------------
Mean:                 25.890067 s
Standard deviation:   4.092455 s
Min:                  21.455162 s
Max:                  34.664623 s
Total time:           258.900666 s
```

sotpapathe committed:
Decoupled preprocessing

sotpapathe committed:
Decoupled tracking

sotpapathe committed:
Moved kfusion and bfusion into include dir and fixed some header includes

sotpapathe committed:
Moved SE_FIELD_TYPE to volume_traits.hpp

This makes it easier to use the typedef in other header files.

sotpapathe committed:
Added include guards and fixed includes

sotpapathe committed:
Split rendering into hpp and cpp

sotpapathe committed:
Fixed bugs in rendering implementations.

Fixed OpenMP parallel fors and arguments passed to TOCK macro.

sotpapathe committed:
Code style improvements

sotpapathe committed:
Updated gitignore

sotpapathe committed:
Merge remote-tracking branch 'upstream/devel' into restructuring

sotpapathe commented:
More benchmarks

#### New code (7e48804a1676841920c29523976adc3428675395)

| dataset | noise\_factor | ATE | preprocessing | integration | raycasting | computation |
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_0` | 0.1000 | 0.0074 | 0.0001 | 0.0123 | 0.0133 | 0.0290 |
| `liv_traj_0` | 0.1000 | 0.0073 | 0.0001 | 0.0122 | 0.0133 | 0.0289 |
| `liv_traj_0` | 0.1000 | 0.0074 | 0.0001 | 0.0153 | 0.0166 | 0.0359 |
| `liv_traj_0` | 0.1000 | 0.0072 | 0.0001 | 0.0124 | 0.0135 | 0.0292 |

| dataset | noise\_factor | ATE | preprocessing | integration | raycasting | computation |
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_1` | 0.1000 | 0.0112 | 0.0012 | 0.0374 | 0.0404 | 0.1019 |
| `liv_traj_1` | 0.1000 | 0.0113 | 0.0001 | 0.0093 | 0.0108 | 0.0230 |
| `liv_traj_1` | 0.1000 | 0.0112 | 0.0004 | 0.0374 | 0.0444 | 0.0938 |
| `liv_traj_1` | 0.1000 | 0.0112 | 0.0008 | 0.0137 | 0.0131 | 0.0336 |

| dataset | noise\_factor | ATE | preprocessing | integration | raycasting | computation |
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_2` | 0.1000 | 0.0037 | 0.0003 | 0.0207 | 0.0188 | 0.0446 |
| `liv_traj_2` | 0.1000 | 0.0037 | 0.0002 | 0.0152 | 0.0137 | 0.0329 |
| `liv_traj_2` | 0.1000 | 0.0038 | 0.0002 | 0.0271 | 0.0240 | 0.0565 |
| `liv_traj_2` | 0.1000 | 0.0038 | 0.0005 | 0.0173 | 0.0141 | 0.0365 |

| dataset | noise\_factor | ATE | preprocessing | integration | raycasting | computation |
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_3` | 0.1000 | 0.7451 | 0.0001 | 0.0141 | 0.0156 | 0.0345 |
| `liv_traj_3` | 0.1000 | 0.6402 | 0.0002 | 0.0069 | 0.0118 | 0.0226 |
| `liv_traj_3` | 0.1000 | 0.6139 | 0.0001 | 0.0130 | 0.0157 | 0.0331 |
| `liv_traj_3` | 0.1000 | 1.2634 | 0.0003 | 0.0049 | 0.0092 | 0.0170 |



#### Old code (devel 0a90a2a2dc37a6ec18113dd4e56f0b45770830be)

| dataset | noise\_factor | ATE | preprocessing | integration | raycasting | computation |
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_0` | 0.1000 | 0.0079 | 0.0003 | 0.0183 | 0.0188 | 0.0440 |
| `liv_traj_0` | 0.1000 | 0.0079 | 0.0001 | 0.0163 | 0.0176 | 0.0381 |
| `liv_traj_0` | 0.1000 | 0.0079 | 0.0001 | 0.0153 | 0.0165 | 0.0357 |
| `liv_traj_0` | 0.1000 | 0.0076 | 0.0001 | 0.0145 | 0.0153 | 0.0336 |

| dataset | noise\_factor | ATE | preprocessing | integration | raycasting | computation |
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_1` | 0.1000 | 0.0112 | 0.0007 | 0.0442 | 0.0388 | 0.0967 |
| `liv_traj_1` | 0.1000 | 0.0113 | 0.0004 | 0.0419 | 0.0373 | 0.0895 |
| `liv_traj_1` | 0.1000 | 0.0112 | 0.0005 | 0.0376 | 0.0344 | 0.0834 |
| `liv_traj_1` | 0.1000 | 0.0111 | 0.0001 | 0.0133 | 0.0118 | 0.0282 |

| dataset | noise\_factor | ATE | preprocessing | integration | raycasting | computation |
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_2` | 0.1000 | 0.0036 | 0.0002 | 0.0271 | 0.0209 | 0.0524 |
| `liv_traj_2` | 0.1000 | 0.0036 | 0.0002 | 0.0230 | 0.0178 | 0.0447 |
| `liv_traj_2` | 0.1000 | 0.0036 | 0.0002 | 0.0251 | 0.0195 | 0.0494 |
| `liv_traj_2` | 0.1000 | 0.0036 | 0.0001 | 0.0178 | 0.0138 | 0.0346 |

| dataset | noise\_factor | ATE | preprocessing | integration | raycasting | computation |
| :- | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_3` | 0.1000 | 1.1698 | 0.0001 | 0.0075 | 0.0095 | 0.0206 |
| `liv_traj_3` | 0.1000 | 1.2615 | 0.0002 | 0.0048 | 0.0104 | 0.0182 |
| `liv_traj_3` | 0.1000 | 1.2629 | 0.0001 | 0.0057 | 0.0126 | 0.0207 |
| `liv_traj_3` | 0.1000 | 0.7542 | 0.0002 | 0.0129 | 0.0123 | 0.0293 |

sotpapathe committed:
Merge remote-tracking branch 'upstream/devel' into restructuring

sotpapathe commented:
Merged with the latest devel and added TUM and ofusion benchmarks. I didn't
notice any significant changed between the old and new version. Timings vary
because I was using the system running the benchmark but most of them don't
seem to vary much.

For some reason `liv_traj_3` consistently crashed with a Segmentation Fault
when running the benchmark with ofusion both with the new and old code. The
crash does not happen when running with the GUI and not saving the log and the
volume. I can look into it later.


## SDF

#### New code (e8d03c4098b3b2de4acca3acfa72f84d0dd09d53)

| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_xyz`         | sdf | 0.0182 | 0.0048 | 0.0305 | 0.0220 | 0.1163 | 0.1000 | 512 |
| `fr1_xyz`         | sdf | 0.0182 | 0.0001 | 0.0097 | 0.0160 | 0.0303 | 0.1000 | 512 |
| `fr1_des`         | sdf | 0.0498 | 0.0064 | 0.0484 | 0.0247 | 0.1927 | 0.1000 | 512 |
| `fr1_des`         | sdf | 0.0499 | 0.0002 | 0.0107 | 0.0138 | 0.0305 | 0.1000 | 512 |
| `fr2_des`         | sdf | 0.2827 | 0.0003 | 0.0179 | 0.0145 | 0.0388 | 0.1000 | 512 |
| `fr2_des`         | sdf | 0.2820 | 0.0001 | 0.0165 | 0.0140 | 0.0348 | 0.1000 | 512 |
| `fr3_long_office` | sdf | 0.6147 | 0.0002 | 0.0197 | 0.0139 | 0.0383 | 0.1000 | 512 |
| `fr3_long_office` | sdf | 0.6169 | 0.0002 | 0.0253 | 0.0181 | 0.0495 | 0.1000 | 512 |



#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)

| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_xyz`         | sdf | 0.0182 | 0.0002 | 0.0091 | 0.0143 | 0.0278 | 0.1000 | 512 |
| `fr1_xyz`         | sdf | 0.0182 | 0.0006 | 0.0098 | 0.0147 | 0.0314 | 0.1000 | 512 |
| `fr1_desk`        | sdf | 0.0501 | 0.0001 | 0.0103 | 0.0134 | 0.0283 | 0.1000 | 512 |
| `fr1_desk`        | sdf | 0.0500 | 0.0001 | 0.0095 | 0.0126 | 0.0268 | 0.1000 | 512 |
| `fr2_desk`        | sdf | 0.2863 | 0.0002 | 0.0170 | 0.0144 | 0.0360 | 0.1000 | 512 |
| `fr2_desk`        | sdf | 0.2847 | 0.0001 | 0.0161 | 0.0138 | 0.0339 | 0.1000 | 512 |
| `fr3_long_office` | sdf | 0.6118 | 0.0001 | 0.0180 | 0.0128 | 0.0348 | 0.1000 | 512 |
| `fr3_long_office` | sdf | 0.6153 | 0.0001 | 0.0201 | 0.0142 | 0.0386 | 0.1000 | 512 |



## OFUSION

#### New code (e8d03c4098b3b2de4acca3acfa72f84d0dd09d53)

| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_0`      | ofusion | 0.0091 | 0.0002 | 0.0506 | 0.0411 | 0.0955 | 0.1000 | 512 |
| `liv_traj_0`      | ofusion | 0.0092 | 0.0004 | 0.0916 | 0.0711 | 0.1705 | 0.1000 | 512 |
| `liv_traj_1`      | ofusion | 0.0242 | 0.0002 | 0.0395 | 0.0365 | 0.0796 | 0.1000 | 512 |
| `liv_traj_1`      | ofusion | 0.0230 | 0.0001 | 0.0431 | 0.0413 | 0.0877 | 0.1000 | 512 |
| `liv_traj_2`      | ofusion | 0.0075 | 0.0001 | 0.0502 | 0.0396 | 0.0925 | 0.1000 | 512 |
| `liv_traj_2`      | ofusion | 0.0074 | 0.0002 | 0.0505 | 0.0391 | 0.0926 | 0.1000 | 512 |
| `fr1_xyz`         | ofusion | 0.0184 | 0.0001 | 0.0249 | 0.0436 | 0.0717 | 0.1000 | 512 |
| `fr1_xyz`         | ofusion | 0.0184 | 0.0001 | 0.0246 | 0.0435 | 0.0714 | 0.1000 | 512 |
| `fr1_desk`        | ofusion | 0.0324 | 0.0001 | 0.0278 | 0.0472 | 0.0785 | 0.1000 | 512 |
| `fr1_desk`        | ofusion | 0.0325 | 0.0001 | 0.0280 | 0.0467 | 0.0784 | 0.1000 | 512 |
| `fr2_desk`        | ofusion | 0.1271 | 0.0001 | 0.0449 | 0.0492 | 0.0983 | 0.1000 | 512 |
| `fr2_desk`        | ofusion | 0.1260 | 0.0001 | 0.0449 | 0.0493 | 0.0985 | 0.1000 | 512 |
| `fr3_long_office` | ofusion | 0.0346 | 0.0001 | 0.0460 | 0.0442 | 0.0937 | 0.1000 | 512 |
| `fr3_long_office` | ofusion | 0.0346 | 0.0001 | 0.0514 | 0.0494 | 0.1047 | 0.1000 | 512 |



#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)

| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_0`      | ofusion | 0.0095 | 0.0022 | 0.0653| 0.0415 | 0.1223 | 0.1000 | 512 |
| `liv_traj_0`      | ofusion | 0.0092 | 0.0001 | 0.0433| 0.0345 | 0.0811 | 0.1000 | 512 |
| `liv_traj_1`      | ofusion | 0.0238 | 0.0001 | 0.0371| 0.0352 | 0.0750 | 0.1000 | 512 |
| `liv_traj_1`      | ofusion | 0.0237 | 0.0001 | 0.0379| 0.0356 | 0.0764 | 0.1000 | 512 |
| `liv_traj_2`      | ofusion | 0.0075 | 0.0002 | 0.0489| 0.0381 | 0.0922 | 0.1000 | 512 |
| `liv_traj_2`      | ofusion | 0.0075 | 0.0001 | 0.0491| 0.0384 | 0.0901 | 0.1000 | 512 |
| `fr1_xyz`         | ofusion | 0.0184 | 0.0002 | 0.0291| 0.0489 | 0.0825 | 0.1000 | 512 |
| `fr1_xyz`         | ofusion | 0.0184 | 0.0001 | 0.0251| 0.0442 | 0.0725 | 0.1000 | 512 |
| `fr1_desk`        | ofusion | 0.0325 | 0.0003 | 0.0889| 0.1527 | 0.2629 | 0.1000 | 512 |
| `fr1_desk`        | ofusion | 0.0324 | 0.0001 | 0.0279| 0.0471 | 0.0786 | 0.1000 | 512 |
| `fr2_desk`        | ofusion | 0.1268 | 0.0001 | 0.0620| 0.0664 | 0.1337 | 0.1000 | 512 |
| `fr2_desk`        | ofusion | 0.1270 | 0.0001 | 0.0456| 0.0495 | 0.0989 | 0.1000 | 512 |
| `fr3_long_office` | ofusion | 0.0346 | 0.0002 | 0.0531| 0.0515 | 0.1092 | 0.1000 | 512 |
| `fr3_long_office` | ofusion | 0.0346 | 0.0002 | 0.0519| 0.0486 | 0.1050 | 0.1000 | 512 |

andreanicastro commented on se_denseslam/include/se/rendering.hpp:
```diff
@@ -0,0 +1,202 @@
+  TOCK("raycastKernel", vertex.width() * vertex.height());
+}
+
+// void renderNormalKernel(uchar3* out, const float3* normal, uint2 normalSize) {
```
Since we are here, why don't you remove commented code? 
Thanks!

sotpapathe replied:
Removed commented out code in rendering.hpp and tracking.cpp

andreanicastro commented on se_denseslam/include/se/rendering.hpp:
```diff
@@ -0,0 +1,202 @@
+                        const float                       largestep,
+                        const Eigen::Vector3f&            light,
+                        const Eigen::Vector3f&            ambient,
+                        bool                              render,
```
const?

sotpapathe replied:
Made const along with some other function arguments.

andreanicastro commented on se_denseslam/src/preprocessing.cpp:
```diff
@@ -31,16 +31,16 @@
-		const std::vector<float>& gaussian, float e_d, int r) {
+void bilateralFilterKernel(se::Image<float>&         out,
+                           const se::Image<float>&   in,
+		                   const std::vector<float>& gaussian,
```
indent

sotpapathe replied:
Fixed indentation and converted tabs to spaces where needed.

andreanicastro commented on se_denseslam/src/preprocessing.cpp:
```diff
@@ -53,7 +53,7 @@ void bilateralFilterKernel(se::Image<float>& out, const se::Image<float>& in,
 		int y;
 		float e_d_squared_2 = e_d * e_d * 2;
```
const?

sotpapathe replied:
Made const.

andreanicastro commented on se_denseslam/src/tracking.cpp:
```diff
@@ -106,7 +105,7 @@ void new_reduce(int blockIndex, float * out, TrackData* J,
 	sums30 = 0.0f;
 	sums31 = 0.0f;
-// comment me out to try coarse grain parallelism 
+// comment me out to try coarse grain parallelism
```
Remove this comment

sotpapathe replied:
Removed.

andreanicastro commented on se_denseslam/include/se/rendering.hpp:
```diff
@@ -0,0 +1,202 @@
+      Eigen::Vector3f test, surfNorm;
+      const int idx = (x + depthSize.x()*y) * 4;
+
+      if (render) {
```
Since we are cleaning this up... I am not sure I like this if here, but why don't we call this variable something like `raycast_normals` instead of `render`?

sotpapathe replied:
Renamed to raycast_normals and also renamed a few other variables.

andreanicastro commented on se_denseslam/include/se/tracking.hpp:
```diff
@@ -0,0 +1,78 @@
+
+bool updatePoseKernel(Eigen::Matrix4f& pose,
+                      const float*     output,
+		              const float      icp_threshold);
```
indentation

sotpapathe replied:
Fixed

andreanicastro commented on se_denseslam/src/preprocessing.cpp:
```diff
@@ -31,16 +31,16 @@
+                           const se::Image<float>&   in,
+		                   const std::vector<float>& gaussian,
+                           const float               e_d,
+                           const int                 r) {
```
avoid single letter variables as they are hard to search... is this the radius of the kernel? `radius`?

sotpapathe replied:
Renamed to radius.

andreanicastro commented on se_denseslam/src/preprocessing.cpp:
```diff
@@ -208,9 +215,9 @@ void halfSampleRobustImageKernel(se::Image<float>& out,
-                Eigen::Vector2i::Constant(0), 
+          Eigen::Vector2i cur = centerPixel + Eigen::Vector2i(j, i);
+          se::math::clamp(cur,
+                Eigen::Vector2i::Constant(0),
```
`Eigen::Vector2i::Zero()`?  it is slightly more readable

sotpapathe replied:
Changed Constant() to Zero()/Ones() where appropriate.

andreanicastro reviewed (changes requested):
Overall the PR looks OK to me. The comments I made are mostly to improve readability. Since you touched that code it fell on you the duty of fixing this small issues, sorry about that!

sotpapathe committed:
Code quality improvements and parallel for fixes

Added missing const.
Changed variable names from camel case to underscores.
Removed commented out code.
Fixed indentation and converted tabs to spaces.
Changed some variable names.
Changed Eigen Constant() to Zero() or Ones() where appropriate.

sotpapathe commented:
Thanks for the comments @andreanicastro !

- Added several missing const in function and variable declarations.
- Fixed some OpenMP parallel fors.
- Removed commented out code.
- Fixed indentation and converted tabs to spaces.
- Changed Eigen Constant() to Zero() or Ones() where appropriate.
- Changed a few variable names to be more descriptive.
- Changed function argument names from camel case to underscores.

andreanicastro commented:
Thanks @sotpapathe ! 
It looks good to me. I was just having a look at the benchmarks and looks like some of the ATEs had a bit of regression. Before merging I would like to hear @emanuelev 's opinion on this

sotpapathe commented:
In certain sequences e.g. `liv_traj_3` tracking is not good at some frames resulting in large variations in the ATE. As a rule of thumb, sequences where the first decimal digit of the ATE is 0 e.g. 0.0XXX are quite consistent, whereas sequences where it is not zero indicate difficulty in tracking and the results vary more. I'd also like to hear @emanuelev 's opinion though.

emanuelev commented:
@sotpapathe Can you please also post the execution platform, OS, compiler version and whether OpenMP was turned on or off? Many thanks!

sotpapathe commented:
All benchmarks were run on:
- Ubuntu 18.04
- Linux kernel 4.15
- gcc 7.3.0
- OpenMP on
- i7-8750H @ 2.20GHz (6 cores - 12 threads) and 8GB RAM

New OFUSION benchmarks with smaller noise factor (0.01 and 0.005). It seems
that the ATE of `liv_traj_1` has gotten worse in the 3rd decimal digit.



## OFUSION Noise factor 0.01

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_0`      | ofusion | 0.0158 | 0.0002 | 0.0143 | 0.0232 | 0.0505 | 0.0100 | 512 |
| `liv_traj_0`      | ofusion | 0.0196 | 0.0007 | 0.0282 | 0.0444 | 0.1376 | 0.0100 | 512 |
| `liv_traj_0`      | ofusion | 0.0228 | 0.0002 | 0.0135 | 0.0224 | 0.0491 | 0.0100 | 512 |
| `liv_traj_0`      | ofusion | 0.4961 | 0.0005 | 0.0247 | 0.0479 | 0.1196 | 0.0100 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_0`       | ofusion | 0.0484 | 0.0002 | 0.0240 | 0.0445 | 0.0755 | 0.0100 | 512 |
| `liv_traj_0`       | ofusion | 0.1199 | 0.0004 | 0.0264 | 0.0472 | 0.0839 | 0.0100 | 512 |
| `liv_traj_0`       | ofusion | 0.2687 | 0.0003 | 0.0266 | 0.0500 | 0.1100 | 0.0100 | 512 |
| `liv_traj_0`       | ofusion | 0.6165 | 0.0001 | 0.0110 | 0.0235 | 0.0376 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_1`      | ofusion | 0.0218 | 0.0001 | 0.0144 | 0.0221 | 0.0528 | 0.0100 | 512 |
| `liv_traj_1`      | ofusion | 0.0218 | 0.0001 | 0.0186 | 0.0292 | 0.0665 | 0.0100 | 512 |
| `liv_traj_1`      | ofusion | 0.0219 | 0.0002 | 0.0149 | 0.0237 | 0.0540 | 0.0100 | 512 |
| `liv_traj_1`      | ofusion | 0.0220 | 0.0001 | 0.0198 | 0.0316 | 0.0733 | 0.0100 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_1`       | ofusion | 0.0189 | 0.0001 | 0.0209 | 0.0352 | 0.0606 | 0.0100 | 512 |
| `liv_traj_1`       | ofusion | 0.0190 | 0.0002 | 0.0240 | 0.0423 | 0.0723 | 0.0100 | 512 |
| `liv_traj_1`       | ofusion | 0.0191 | 0.0001 | 0.0134 | 0.0222 | 0.0385 | 0.0100 | 512 |
| `liv_traj_1`       | ofusion | 0.0191 | 0.0002 | 0.0294 | 0.0510 | 0.0871 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_2`      | ofusion | 0.0079 | 0.0002 | 0.0213 | 0.0260 | 0.0608 | 0.0100 | 512 |
| `liv_traj_2`      | ofusion | 0.0079 | 0.0003 | 0.0229 | 0.0266 | 0.0630 | 0.0100 | 512 |
| `liv_traj_2`      | ofusion | 0.0079 | 0.0004 | 0.0250 | 0.0291 | 0.0763 | 0.0100 | 512 |
| `liv_traj_2`      | ofusion | 0.0079 | 0.0005 | 0.0249 | 0.0282 | 0.0713 | 0.0100 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_2`       | ofusion | 0.0079 | 0.0001 | 0.0194 | 0.0241 | 0.0462 | 0.0100 | 512 |
| `liv_traj_2`       | ofusion | 0.0079 | 0.0001 | 0.0241 | 0.0301 | 0.0579 | 0.0100 | 512 |
| `liv_traj_2`       | ofusion | 0.0079 | 0.0001 | 0.0252 | 0.0312 | 0.0598 | 0.0100 | 512 |
| `liv_traj_2`       | ofusion | 0.0079 | 0.0003 | 0.0240 | 0.0276 | 0.0678 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_3`      | ofusion | 0.0797 | 0.0002 | 0.0152 | 0.0210 | 0.0638 | 0.0100 | 512 |
| `liv_traj_3`      | ofusion | 0.0933 | 0.0002 | 0.0163 | 0.0235 | 0.0601 | 0.0100 | 512 |
| `liv_traj_3`      | ofusion | 0.1304 | 0.0002 | 0.0149 | 0.0223 | 0.0547 | 0.0100 | 512 |
| `liv_traj_3`      | ofusion | 0.1309 | 0.0001 | 0.0134 | 0.0219 | 0.0524 | 0.0100 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_3`       | ofusion | 0.0713 | 0.0002 | 0.0148 | 0.0225 | 0.0438 | 0.0100 | 512 |
| `liv_traj_3`       | ofusion | 0.0749 | 0.0001 | 0.0147 | 0.0230 | 0.0415 | 0.0100 | 512 |
| `liv_traj_3`       | ofusion | 0.0919 | 0.0001 | 0.0167 | 0.0249 | 0.0456 | 0.0100 | 512 |
| `liv_traj_3`       | ofusion | 0.1401 | 0.0001 | 0.0122 | 0.0211 | 0.0363 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_xyz`         | ofusion | 0.0185 | 0.0001 | 0.0078 | 0.0282 | 0.0539 | 0.0100 | 512 |
| `fr1_xyz`         | ofusion | 0.0185 | 0.0001 | 0.0080 | 0.0269 | 0.0580 | 0.0100 | 512 |
| `fr1_xyz`         | ofusion | 0.0185 | 0.0002 | 0.0086 | 0.0297 | 0.0568 | 0.0100 | 512 |
| `fr1_xyz`         | ofusion | 0.0185 | 0.0005 | 0.0092 | 0.0296 | 0.0633 | 0.0100 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_xyz`          | ofusion | 0.0185 | 0.0001 | 0.0080 | 0.0272 | 0.0386 | 0.0100 | 512 |
| `fr1_xyz`          | ofusion | 0.0185 | 0.0001 | 0.0086 | 0.0295 | 0.0418 | 0.0100 | 512 |
| `fr1_xyz`          | ofusion | 0.0185 | 0.0001 | 0.0089 | 0.0305 | 0.0432 | 0.0100 | 512 |
| `fr1_xyz`          | ofusion | 0.0185 | 0.0003 | 0.0108 | 0.0299 | 0.0605 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_desk`        | ofusion | 0.0485 | 0.0001 | 0.0082 | 0.0287 | 0.0574 | 0.0100 | 512 |
| `fr1_desk`        | ofusion | 0.0487 | 0.0002 | 0.0085 | 0.0290 | 0.0573 | 0.0100 | 512 |
| `fr1_desk`        | ofusion | 0.0491 | 0.0004 | 0.0103 | 0.0298 | 0.0744 | 0.0100 | 512 |
| `fr1_desk`        | ofusion | 0.0491 | 0.0005 | 0.0092 | 0.0299 | 0.0641 | 0.0100 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_desk`         | ofusion | 0.0475 | 0.0001 | 0.0091 | 0.0296 | 0.0427 | 0.0100 | 512 |
| `fr1_desk`         | ofusion | 0.0477 | 0.0003 | 0.0098 | 0.0296 | 0.0534 | 0.0100 | 512 |
| `fr1_desk`         | ofusion | 0.0505 | 0.0001 | 0.0093 | 0.0307 | 0.0441 | 0.0100 | 512 |
| `fr1_desk`         | ofusion | 0.0516 | 0.0001 | 0.0086 | 0.0287 | 0.0410 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr2_desk`        | ofusion | 0.1085 | 0.0001 | 0.0188 | 0.0306 | 0.0685 | 0.0100 | 512 |
| `fr2_desk`        | ofusion | 0.1085 | 0.0001 | 0.0191 | 0.0314 | 0.0685 | 0.0100 | 512 |
| `fr2_desk`        | ofusion | 0.1086 | 0.0003 | 0.0207 | 0.0324 | 0.0742 | 0.0100 | 512 |
| `fr2_desk`        | ofusion | 0.1086 | 0.0003 | 0.0227 | 0.0364 | 0.0805 | 0.0100 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr2_desk`         | ofusion | 0.1078 | 0.0003 | 0.0240 | 0.0370 | 0.0717 | 0.0100 | 512 |
| `fr2_desk`         | ofusion | 0.1078 | 0.0004 | 0.0240 | 0.0315 | 0.1042 | 0.0100 | 512 |
| `fr2_desk`         | ofusion | 0.1079 | 0.0001 | 0.0193 | 0.0308 | 0.0536 | 0.0100 | 512 |
| `fr2_desk`         | ofusion | 0.1079 | 0.0001 | 0.0198 | 0.0310 | 0.0550 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr3_long_office` | ofusion | 0.0316 | 0.0002 | 0.0209 | 0.0305 | 0.0706 | 0.0100 | 512 |
| `fr3_long_office` | ofusion | 0.0317 | 0.0001 | 0.0222 | 0.0330 | 0.0743 | 0.0100 | 512 |
| `fr3_long_office` | ofusion | 0.0317 | 0.0003 | 0.0217 | 0.0307 | 0.0741 | 0.0100 | 512 |
| `fr3_long_office` | ofusion | 0.0317 | 0.0003 | 0.0335 | 0.0491 | 0.1175 | 0.0100 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr3_long_office`  | ofusion | 0.0311 | 0.0001 | 0.0203 | 0.0288 | 0.0526 | 0.0100 | 512 |
| `fr3_long_office`  | ofusion | 0.0311 | 0.0001 | 0.0205 | 0.0289 | 0.0531 | 0.0100 | 512 |
| `fr3_long_office`  | ofusion | 0.0311 | 0.0001 | 0.0228 | 0.0326 | 0.0594 | 0.0100 | 512 |
| `fr3_long_office`  | ofusion | 0.0311 | 0.0003 | 0.0421 | 0.0595 | 0.1178 | 0.0100 | 512 |



## OFUSION Noise factor 0.005

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_0`      | ofusion | 0.3423 | 0.0001 | 0.0118 | 0.0227 | 0.0491 | 0.0050 | 512 |
| `liv_traj_0`      | ofusion | 0.3691 | 0.0005 | 0.0254 | 0.0491 | 0.1216 | 0.0050 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_0`      | ofusion | 0.1199 | 0.0004 | 0.0264 | 0.0472 | 0.0839 | 0.0100 | 512 |
| `liv_traj_0`      | ofusion | 0.6008 | 0.0012 | 0.0225 | 0.0312 | 0.0644 | 0.0050 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_1`      | ofusion | 0.0183 | 0.0002 | 0.0132 | 0.0206 | 0.0550 | 0.0050 | 512 |
| `liv_traj_1`      | ofusion | 0.0184 | 0.0003 | 0.0170 | 0.0301 | 0.0688 | 0.0050 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_1`      | ofusion | 0.0190 | 0.0002 | 0.0240 | 0.0423 | 0.0723 | 0.0100 | 512 |
| `liv_traj_1`      | ofusion | 0.0332 | 0.0008 | 0.0141 | 0.0248 | 0.0473 | 0.0050 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_2`      | ofusion | 0.0079 | 0.0002 | 0.0212 | 0.0233 | 0.0651 | 0.0050 | 512 |
| `liv_traj_2`      | ofusion | 0.0080 | 0.0001 | 0.0238 | 0.0292 | 0.0697 | 0.0050 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_2`      | ofusion | 0.0079 | 0.0001 | 0.0241 | 0.0301 | 0.0579 | 0.0100 | 512 |
| `liv_traj_2`      | ofusion | 0.0080 | 0.0008 | 0.0273 | 0.0295 | 0.0654 | 0.0050 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_3`      | ofusion | 0.0956 | 0.0002 | 0.0138 | 0.0217 | 0.0552 | 0.0050 | 512 |
| `liv_traj_3`      | ofusion | 0.1169 | 0.0001 | 0.0126 | 0.0197 | 0.0499 | 0.0050 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `liv_traj_3`      | ofusion | 0.0680 | 0.0007 | 0.0188 | 0.0261 | 0.0545 | 0.0050 | 512 |
| `liv_traj_3`      | ofusion | 0.0919 | 0.0001 | 0.0167 | 0.0249 | 0.0456 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_xyz`         | ofusion | 0.0185 | 0.0001 | 0.0069 | 0.0262 | 0.0505 | 0.0050 | 512 |
| `fr1_xyz`         | ofusion | 0.0185 | 0.0003 | 0.0095 | 0.0328 | 0.0687 | 0.0050 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_xyz`         | ofusion | 0.0185 | 0.0001 | 0.0089 | 0.0305 | 0.0432 | 0.0100 | 512 |
| `fr1_xyz`         | ofusion | 0.0185 | 0.0012 | 0.0147 | 0.0335 | 0.0598 | 0.0050 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_desk`        | ofusion | 0.0467 | 0.0002 | 0.0084 | 0.0299 | 0.0605 | 0.0050 | 512 |
| `fr1_desk`        | ofusion | 0.0472 | 0.0003 | 0.0083 | 0.0282 | 0.0573 | 0.0050 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr1_desk`        | ofusion | 0.0501 | 0.0009 | 0.0137 | 0.0334 | 0.0580 | 0.0050 | 512 |
| `fr1_desk`        | ofusion | 0.0505 | 0.0001 | 0.0093 | 0.0307 | 0.0441 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr2_desk`        | ofusion | 0.1066 | 0.0002 | 0.0187 | 0.0319 | 0.0720 | 0.0050 | 512 |
| `fr2_desk`        | ofusion | 0.1067 | 0.0002 | 0.0209 | 0.0359 | 0.0776 | 0.0050 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr2_desk`        | ofusion | 0.1056 | 0.0015 | 0.0291 | 0.0381 | 0.0879 | 0.0050 | 512 |
| `fr2_desk`        | ofusion | 0.1079 | 0.0001 | 0.0198 | 0.0310 | 0.0550 | 0.0100 | 512 |

#### New code (ab3a4c9b152708a78dab81e2b730cae0cd17d6c8)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr3_long_office` | ofusion | 0.0301 | 0.0001 | 0.0243 | 0.0398 | 0.0862 | 0.0050 | 512 |
| `fr3_long_office` | ofusion | 0.0302 | 0.0001 | 0.0206 | 0.0318 | 0.0724 | 0.0050 | 512 |

#### Old code (devel fbbbd0f0ed90ed9e0fd9413eff51cef9d9bac795)
| dataset | implementation | ATE | preprocessing | integration | raycasting | computation | noise\_factor | resolution |
| :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| `fr3_long_office` | ofusion | 0.0299 | 0.0014 | 0.0308 | 0.0374 | 0.0823 | 0.0050 | 512 |
| `fr3_long_office` | ofusion | 0.0311 | 0.0003 | 0.0421 | 0.0595 | 0.1178 | 0.0100 | 512 |
